File: DSA/Pranvkumar_590011587/Day40_Question1.py
def findCelebrity(M):
    n = len(M)
    stack = list(range(n))
    # Eliminate non-celebrities
    while len(stack) > 1:
        a = stack.pop()
        b = stack.pop()
        if M[a][b]:
            stack.append(b)
        else:
            stack.append(a)
    if not stack:
        return -1
    candidate = stack.pop()
    # Verify candidate
    for i in range(n):
        if i != candidate and (M[candidate][i] == 1 or M[i][candidate] == 0):
            return -1
    return candidate

# Examples:
print(findCelebrity([[0, 1, 1], [0, 0, 1], [0, 0, 0]])) 
print(findCelebrity([[0, 1], [1, 0]]))                 

File: DSA/VenikaPanwar_590013755/Day_40_Question_1.java
import java.util.Scanner;
import java.util.Stack;

public class Day_40_Question_1 {

    static boolean knows(int[][] matrix, int a, int b) {
        return matrix[a][b] == 1;
    }

    static int findCelebrity(int[][] matrix, int n) {
        Stack<Integer> stack = new Stack<>();

        // Push all people into stack
        for (int i = 0; i < n; i++) stack.push(i);

        // Eliminate non-celebrities
        while (stack.size() > 1) {
            int a = stack.pop();
            int b = stack.pop();

            if (knows(matrix, a, b)) {
                // a cannot be celebrity
                stack.push(b);
            } else {
                // b cannot be celebrity
                stack.push(a);
            }
        }

        // Final candidate
        int candidate = stack.pop();

        // Validate candidate
        for (int i = 0; i < n; i++) {
            if (i != candidate && (matrix[candidate][i] == 1 || matrix[i][candidate] == 0)) {
                return -1;
            }
        }

        return candidate;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter number of people: ");
        int n = sc.nextInt();

        int[][] matrix = new int[n][n];
        System.out.println("Enter matrix:");
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                matrix[i][j] = sc.nextInt();

        int result = findCelebrity(matrix, n);
        System.out.println("Celebrity Index: " + result);
    }
}
